packages/ifpack2/src/Ifpack2_BlockTriDiContainer_def.hpp

```diff
@@ -0,0 +1,2678 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+//@HEADER
+*/
+
+#ifndef IFPACK2_BLOCKTRIDICONTAINER_DEF_HPP
+#define IFPACK2_BLOCKTRIDICONTAINER_DEF_HPP
+
+#include <Teuchos_Details_MpiTypeTraits.hpp>
+
+#include <Tpetra_Distributor.hpp>
+#include <Tpetra_BlockMultiVector.hpp>
+
+#include <Kokkos_ArithTraits.hpp>
+#include <KokkosBatched_Vector.hpp>
+#include <KokkosBatched_Gemm_Decl.hpp>
+#include <KokkosBatched_Gemm_Serial_Impl.hpp>
+#include <KokkosBatched_Gemv_Decl.hpp>
+#include <KokkosBatched_Gemv_Serial_Impl.hpp>
+#include <KokkosBatched_Trsm_Decl.hpp>
+#include <KokkosBatched_Trsm_Serial_Impl.hpp>
+#include <KokkosBatched_Trsv_Decl.hpp>
+#include <KokkosBatched_Trsv_Serial_Impl.hpp>
+#include <KokkosBatched_LU_Decl.hpp>
+#include <KokkosBatched_LU_Serial_Impl.hpp>
+
+#include "Ifpack2_BlockTriDiContainer_decl.hpp"
+
+//todo Move to KokkosKernels.
+#if defined(__AVX__) || defined(__AVX2__) || defined(__AVX512F__)
+# include <immintrin.h>
+#endif
+
+namespace KokkosBatched {
+namespace Details {
+namespace Todo {
+using KokkosBatched::Experimental::Vector;
+using KokkosBatched::Experimental::VectorTag;
+using KokkosBatched::Experimental::AVX;
+
+template <typename ST1, typename ST2> KOKKOS_INLINE_FUNCTION ST1
+add_radial (const ST1& a, const ST2& r) { return a >= 0 ? a+r : a-r; }
+
+#ifdef HAVE_TEUCHOS_COMPLEX
+template <typename ST1, typename ST2> KOKKOS_INLINE_FUNCTION std::complex<ST1>
+add_radial (const std::complex<ST1>& a, const ST2& r) {
+  const auto a_abs = std::abs(a);
+  const auto sign = a_abs == Kokkos::ArithTraits<ST1>::zero() ? 1 : a/a_abs;
+  return sign*(a_abs + r);
+}
+#endif
+
+#if defined(__AVX__) || defined(__AVX2__) || defined(__AVX512F__)
+inline static __m256d add_radial_impl (const __m256d& a, const double r) {
+  // Make the mask 10...0.
+  const __m256d mask = _mm256_set1_pd(-0.);
+  // Use it to get the sign.
+  const __m256d sign = _mm256_and_pd(mask, a);
+  // Load r.
+  const __m256d pr = _mm256_set1_pd(r);
+  // Now use the opposite mask, 01...1, to get b = abs(a).
+  __m256d pb = _mm256_andnot_pd(mask, a);
+  // Compute abs(a) + r.
+  pb = _mm256_add_pd(pb, pr);
+  // Bit-or the sign back in.
+  pb = _mm256_or_pd(pb, sign);
+  return pb;
+}
+#endif
```

Comment:
@kyungjoo-kim, I'm mimicking what the application does, which is a requirement. I agree with what you write (but watch out for the complex case, where the complex sign is needed), so feel free to put that in KokkosBatched. If the application doesn't want to switch to that (eventually), then we'll keep the special impl for the app in BlockTriDiContainer.